File: 백준/Silver/1149. RGB거리/RGB거리.cc
#include <iostream>
#include <algorithm>
using namespace std;

const int MAX = 1000;
int	d[MAX + 1][3];
int a[MAX + 1][3];

int main(void) {
	ios_base::sync_with_stdio(false);
	cin.tie(nullptr);
	cout.tie(nullptr);

	int N;
	cin >> N;	//input num of house

	for (int i = 1; i <= N; i++) {		//input color cost of RGB for each house
		for (int j = 0; j < 3; j++) {
			cin>>a[i][j];
		}
	}


	
	for (int i = 1; i <= N; i++) {
			d[i][0] = min(d[i - 1][1], d[i - 1][2]) + a[i][0];
			d[i][1] = min(d[i - 1][0], d[i - 1][2]) + a[i][1];
			d[i][2] = min(d[i - 1][0], d[i - 1][1]) + a[i][2];
	}
	cout << min({ d[N][0] , d[N][1] ,d[N][2] }) << '\n';

	return 0;
}

File: 백준/Silver/1149. RGB거리/RGB거리.swift
import Foundation

let n = Int(readLine()!)!

var cost = Array(repeating: [0,0,0], count: n)
// cost[i]: i번째집을 r, g, b 로 칠하는 비용

for i in 0..<n {
  let input = readLine()!.split(separator: " ").map { Int(String($0))! }
  cost[i] = input
}

var dp = Array(repeating: [0, 0, 0], count: n)

// dp[i][k] i번째집을 r: 0, g: 1, b: 2 로 칠했을때의 비용의 최솟값

dp[0] = cost[0]

for i in 1..<n {
  dp[i][0] = min(dp[i-1][1], dp[i-1][2]) + cost[i][0]
  dp[i][1] = min(dp[i-1][0], dp[i-1][2]) + cost[i][1]
  dp[i][2] = min(dp[i-1][0], dp[i-1][1]) + cost[i][2]
}

print(dp[n-1].min()!)